File: src/com/dafttech/terra/TerraInfinita.java
package com.dafttech.terra;

import java.util.Random;

import org.lwjgl.opengl.Display;

import com.badlogic.gdx.ApplicationListener;
import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.FPSLogger;
import com.badlogic.gdx.graphics.Texture;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.engine.gui.MouseSlot;
import com.dafttech.terra.engine.gui.Tooltip;
import com.dafttech.terra.engine.input.InputHandler;
import com.dafttech.terra.game.Events;
import com.dafttech.terra.game.ScreenIngame;
import com.dafttech.terra.game.ScreenPauseMenu;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.resources.Resources;

public class TerraInfinita extends Game implements ApplicationListener {
    public static TerraInfinita $ = new TerraInfinita();

    FPSLogger fpsLogger;

    public static Random rnd = new Random();
    public ScreenIngame screenIngame;
    public ScreenPauseMenu screenPause;

    public World world;

    boolean wasFocused = false;

    public boolean isFocused() {
        return Display.isActive();
    }

    @Override
    public void create() {

        Gdx.app.log(Thread.currentThread().getName(), "Creating game...");

        Texture.setEnforcePotImages(false);

        Events.init();
        Events.EVENTMANAGER.callSync(Events.EVENT_INITPRE, this);

        fpsLogger = new FPSLogger();

        Resources.init();
        InputHandler.init();
        Tooltip.init();
        MouseSlot.init();

        world = new World(new Vector2(2000, 1000));

        screenIngame = new ScreenIngame(world);
        screenPause = new ScreenPauseMenu(world);

        setScreen(screenPause);

        Events.EVENTMANAGER.callSync(Events.EVENT_INITPOST, this);
    }

    @Override
    public void render() {
        if (!isFocused() && wasFocused) {
            setScreen(screenPause);
            wasFocused = false;
        }

        if (isFocused() && !wasFocused) {
            wasFocused = true;
        }

        super.render();

        fpsLogger.log();
    }

    @Override
    public void resize(int arg0, int arg1) {
        Events.EVENTMANAGER.callSync(Events.EVENT_WINRESIZE, this);
    }

    @Override
    public void pause() {
        Events.EVENTMANAGER.callSync(Events.EVENT_WINPAUSE, this);
    }

    @Override
    public void resume() {
        Events.EVENTMANAGER.callSync(Events.EVENT_WINRESUME, this);
    }

    @Override
    public void dispose() {
        Events.EVENTMANAGER.callSync(Events.EVENT_WINDISPOSE, this);
    }
}

File: terra/src/main/scala/com/dafttech/terra/TerraInfinita.scala
package com.dafttech.terra

import java.util.Random

import com.badlogic.gdx.{ApplicationListener, Game, Gdx}
import com.dafttech.terra.engine.gui.{MouseSlot, Tooltip}
import com.dafttech.terra.engine.input.InputHandler
import com.dafttech.terra.engine.vector.Vector2i
import com.dafttech.terra.game.world.World
import com.dafttech.terra.game.{Events, ScreenIngame, ScreenPauseMenu}
import com.dafttech.terra.resources.Resources
import org.lwjgl.opengl.Display

object TerraInfinita extends Game with ApplicationListener {
  val rnd: Random = new Random()

  private[terra] val fpsLogger: FPSLogger = new FPSLogger()
  var screenIngame: ScreenIngame = null
  var screenPause: ScreenPauseMenu = null
  var world: World = null
  private[terra] var wasFocused: Boolean = false

  def isFocused: Boolean = Display.isActive

  def create: Unit = {
    Gdx.app.log(Thread.currentThread.getName, "Creating game...")
    Events.init()
    Events.EVENTMANAGER.callSync(Events.EVENT_INITPRE, this)
    Resources.init()
    InputHandler.init
    Tooltip.init
    MouseSlot.init
    world = new World(Vector2i(2000, 1000))
    screenIngame = new ScreenIngame(world)
    screenPause = new ScreenPauseMenu(world)
    setScreen(screenPause)
    Events.EVENTMANAGER.callSync(Events.EVENT_INITPOST, this)
  }

  override def render(): Unit = {
    if (!isFocused && wasFocused) {
      setScreen(screenPause)
      wasFocused = false
    }
    if (isFocused && !wasFocused) {
      wasFocused = true
    }
    super.render()
    fpsLogger.tick
  }

  override def resize(width: Int, height: Int) {
    super.resize(width, height)
    Gdx.graphics.setWindowedMode(width, height)
    this.getScreen.resize(width, height)
    Events.EVENTMANAGER.callSync(Events.EVENT_WINRESIZE, this)
  }

  override def pause(): Unit = {
    Events.EVENTMANAGER.callSync(Events.EVENT_WINPAUSE, this)
  }

  override def resume(): Unit = {
    Events.EVENTMANAGER.callSync(Events.EVENT_WINRESUME, this)
  }

  override def dispose(): Unit = {
    Events.EVENTMANAGER.callSync(Events.EVENT_WINDISPOSE, this)
  }
}